src/pop/settings.py

```diff
@@ -0,0 +1,96 @@
+# plugin-output-processing
+# Copyright (C) 2024  Centreon
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import sys
+from enum import Enum
+
+from pydantic import BaseModel, Field, ValidationInfo, field_serializer, field_validator
+
+from pop.globals import DEFAULT_ROLE, Language, Provider
+from pop.logger import logger
+from pop.providers import Ollama, OpenAI
+
+# Disable traceback in case of error, cleaner logs especially for REST API
+sys.tracebacklimit = 0
+
+
+providers = {Provider.OPENAI: OpenAI(), Provider.OLLAMA: Ollama()}
+
+
+class Settings(BaseModel):
+
+    provider: Provider | None = Field(default=None, validate_default=True)
+    model: str | None = Field(default=None, validate_default=True)
+    url: str | None = Field(default=None, validate_default=True)
+    temperature: float = 1
+    length: int = 100
+    language: Language = Language.ENGLISH
+    role: str = DEFAULT_ROLE
+
+    @field_serializer("provider", "language")
+    def serialize_enum(self, enum: Enum | None) -> str:
+        """
+        This method is needed to write enum fields in the configuration file.
+        """
+        return enum.value if isinstance(enum, Enum) else None
+
+    @field_validator("provider")
+    @classmethod
+    def check_provider(cls, name: Provider) -> Provider:
+        """
+        Use the provider given by the user, if it's not available try a new one.
+        """
+        while len(providers) > 0:
+            provider = providers.pop(name, None)
+            provider = provider or providers.popitem()[1]
+            provider.fetch()
+            if not provider.available:
+                logger.warning(
+                    f"{provider.name} is not available. Trying another provider."
+                )
+                continue
+            providers[Provider(value=provider.name)] = provider
+            logger.info(f"Using {provider.name} provider.")
+            return provider.name
```

Comment:
I'm not a fan working with a mutable object `providers` which is a global variable furthermore and modifying it (pop item), then reconstruct it (`providers[Provider(value=provider.name)] = provider).
This means that you have no way of knowing what `providers` has just while reading the code outside of this function.

I think it can lead to misunderstanding or hard to find bugs in the future. I would rather use a for loop here:

```suggestion
        for provider_name in [name] + providers.keys():
            if provider_name not in providers:
                continue
            provider = providers[provider_name]
            provider.fetch()
            if provider.available:
                break
            f"{provider_name} is not available. Trying another provider."
        logger.info(f"Using {provider_name} provider.")
        return provider_name
```

Advantage is that `providers` is not modified and the global variable stay as is.
Inconvenient is that `name` can be repeated twice in case it is define but not available and before everything else.

Reply:
I agree with the fact that altering a constant is not a good idea. I fix it, by defining the constant `PROVIDERS` which stored all providers, and the variable `providers` which is a list of providers with the one selected by the user in first position. I can then use a for loop instead of a while loop.